nixos/modules/system/boot/systemd/initrd.nix

```diff
@@ -371,7 +389,12 @@ in {
       "autofs"
       # systemd-cryptenroll
     ] ++ lib.optional cfg.enableTpm2 "tpm-tis"
-    ++ lib.optional (cfg.enableTpm2 && !(pkgs.stdenv.hostPlatform.isRiscV64 || pkgs.stdenv.hostPlatform.isArmv7)) "tpm-crb";
+    ++ lib.optional (cfg.enableTpm2 && !(pkgs.stdenv.hostPlatform.isRiscV64 || pkgs.stdenv.hostPlatform.isArmv7)) "tpm-crb"
+    ++ lib.optional cfg.package.withEfi "efivarfs";
```

Comment:
Shouldn't this be conditional on `cfg.root == "gpt-auto"`? I don't see why this feature should add kernel modules for users who don't enable it. Adding it is problematic for me because it doesn't exist on armv6l.

Reply:
@lopsided98 `efivarfs` is useful for a few features, notably including hibernation. I'd rather disable it on platforms that don't support it than require it to be explicitly enabled.

If `efivarfs` doesn't exist on a platform, then I'm surprised we build systemd with `withEfi = true`. Maybe that's the change we should make?

Reply:
`withEfi` defaults to [`stdenv.hostPlatform.isEfi`](https://github.com/NixOS/nixpkgs/blob/b06025f1533a1e07b6db3e75151caa155d1c7eb3/lib/systems/inspect.nix#L96). In practice, `isEfi` means "platforms where systemd EFI support can be built". There are unlikely to be any armv6l systems that use EFI, so it would be fine to remove it from `isEfi`, but I'm still not super happy with tying `efivarfs` in the initrd to systemd EFI support.

There is no way to override `boot.initrd.availableKernelModules` to remove an item, so we should be careful about what is added unconditionally. If someone wants to disable EFI support in their kernel for whatever reason, they also have to override systemd and rebuild the world. If hibernation needs `efivarsfs`, it should be automatically enabled in that module as well. I actually have a similar objection to `autofs4`, but it has never caused enough problems for me to try fixing it.

Reply:
@lopsided98

1. You don't have to rebuild the world to override systemd. You can just set `systemd.package = pkgs.systemd.override { withEfi = false; };`
2. Hibernation isn't really something we have a module for. It's just something *systemd is able to do*. So simply having `systemctl` means a user is able to attempt to hibernate.

We could have an option like `boot.initrd.systemd.efiSupport` with a default of `cfg.package.withEfi` so that it can be reasonably overridden. I still think this is wrong, however. Ideally, a system that doesn't have EFI should have `systemd.package` set to a systemd override with EFI disabled.

Alternatively, we could have a new option like `boot.initrd.optionalKernelModules` which lists modules that should be included *if the kernel build has them*. This would be helpful in other situations as well. Like the `all-hardware.nix` profile breaks when you build with the RPi kernel, since it disables many of the modules that `all-hardware.nix` has in `availableKernelModules`.

Reply:
>  I actually have a similar objection to autofs, but it has never caused enough problems for me to try fixing it.

`autofs` is required for systemd to boot as far as I am aware.

Reply:
> > I actually have a similar objection to autofs, but it has never caused enough problems for me to try fixing it.
> 
> `autofs` is required for systemd to boot as far as I am aware. as systemd ships with some required `.automount` units

From the docs on [CONFIG_AUTOFS_FS](https://www.kernelconfig.io/config_autofs_fs?arch=x86&kernelversion=6.7.10):

> The automounter is a tool to automatically mount remote file systems on demand. This implementation is partially kernel-based to reduce overhead in the already-mounted case; this is unlike the BSD automounter (amd), which is a pure user space daemon.
>
> To use the automounter you need the user-space tools from <https://www.kernel.org/pub/linux/daemons/autofs/>; you also want to answer Y to "NFS file system support", below.
> 
> To compile this support as a module, choose M here: the module will be called autofs.
> 
> If you are not a part of a fairly large, distributed network or don't have a laptop which needs to dynamically reconfigure to the local network, you probably do not need an automounter, and can say N here. 

I only see [one `.automount` unit that ships with systemd](https://github.com/systemd/systemd/blob/main/units/proc-sys-fs-binfmt_misc.automount) and it looks trivially replaceable.